src/java.base/share/classes/java/util/stream/DoublePipeline.java

```diff
@@ -263,43 +263,41 @@ public final DoubleStream flatMap(DoubleFunction<? extends DoubleStream> mapper)
                 StreamOpFlag.NOT_SORTED | StreamOpFlag.NOT_DISTINCT | StreamOpFlag.NOT_SIZED) {
             @Override
             Sink<Double> opWrapSink(int flags, Sink<Double> sink) {
-                return new Sink.ChainedDouble<>(sink) {
-                    // true if cancellationRequested() has been called
-                    boolean cancellationRequestedCalled;
-
-                    // cache the consumer to avoid creation on every accepted element
-                    DoubleConsumer downstreamAsDouble = downstream::accept;
+                class FlatMap implements Sink.OfDouble, DoublePredicate {
+                    boolean cancel;
+                    private final boolean shorts = isShortCircuitingPipeline();
 
-                    @Override
-                    public void begin(long size) {
-                        downstream.begin(-1);
-                    }
+                    @Override public void begin(long size) { sink.begin(-1); }
+                    @Override public void end() { sink.end(); }
 
                     @Override
-                    public void accept(double t) {
-                        try (DoubleStream result = mapper.apply(t)) {
+                    public void accept(double input) {
+                        try (DoubleStream result = mapper.apply(input)) {
                             if (result != null) {
-                                if (!cancellationRequestedCalled) {
-                                    result.sequential().forEach(downstreamAsDouble);
-                                } else {
-                                    var s = result.sequential().spliterator();
-                                    do {
-                                    } while (!downstream.cancellationRequested() && s.tryAdvance(downstreamAsDouble));
-                                }
+                                if (shorts)
+                                    result.sequential().allMatch(this);
+                                else
+                                    result.sequential().forEach(sink::accept);
```

Comment:
I think that might create a new double consumer instance for every input element. Alternatively you can compute and cache it as a field, replacing `shorts` and use a `null` check.